Src/Newtonsoft.Json/JsonTextReader.Async.cs

```diff
@@ -0,0 +1,1778 @@
+                        }
+
+                        // start of escape sequence
+                        int escapeStartPos = charPos - 1;
+
+                        char currentChar = _chars[charPos];
+                        charPos++;
+
+                        char writeChar;
+
+                        switch (currentChar)
+                        {
+                            case 'b':
+                                writeChar = '\b';
+                                break;
+                            case 't':
+                                writeChar = '\t';
+                                break;
+                            case 'n':
+                                writeChar = '\n';
+                                break;
+                            case 'f':
+                                writeChar = '\f';
+                                break;
+                            case 'r':
+                                writeChar = '\r';
+                                break;
+                            case '\\':
+                                writeChar = '\\';
+                                break;
+                            case '"':
+                            case '\'':
+                            case '/':
+                                writeChar = currentChar;
+                                break;
+                            case 'u':
+                                _charPos = charPos;
+                                writeChar = await ParseUnicodeAsync(cancellationToken).ConfigureAwait(false);
+
+                                if (StringUtils.IsLowSurrogate(writeChar))
+                                {
+                                    // low surrogate with no preceding high surrogate; this char is replaced
+                                    writeChar = UnicodeReplacementChar;
+                                }
+                                else if (StringUtils.IsHighSurrogate(writeChar))
+                                {
+                                    bool anotherHighSurrogate;
+
+                                    // loop for handling situations where there are multiple consecutive high surrogates
+                                    do
+                                    {
+                                        anotherHighSurrogate = false;
+
+                                        // potential start of a surrogate pair
+                                        if (EnsureChars(2, true) && _chars[_charPos] == '\\' && _chars[_charPos + 1] == 'u')
+                                        {
+                                            char highSurrogate = writeChar;
+
+                                            _charPos += 2;
+                                            writeChar = await ParseUnicodeAsync(cancellationToken).ConfigureAwait(false);
+
+                                            if (StringUtils.IsLowSurrogate(writeChar))
```

Comment:
There was a missed point within this that should have been calling into an async method within that. What's left is hence smaller and also hits a lot of locals used elsewhere, so I don't think it's worth it here.